hockeypuck_rock/delete_keys.py

```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python3
+
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Delete keys from the Hockeypuck Postgres database by fingerprint."""
+
+import argparse
+import logging
+import os
+import re
+import sys
+from typing import List
+
+import psycopg2
+
+logger = logging.getLogger(__name__)
+
+
+def get_db_connection() -> psycopg2.extensions.connection | None:
+    """Connect to the Postgres database.
+
+    Returns:
+        psycopg2.extensions.connection: the database connection.
+    """
+    db_password = os.getenv("POSTGRESQL_DB_PASSWORD")
+    db_name = os.getenv("POSTGRESQL_DB_NAME")
+    db_host = os.getenv("POSTGRESQL_DB_HOSTNAME")
+    db_user = os.getenv("POSTGRESQL_DB_USERNAME")
+    dsn = f"dbname={db_name} user={db_user} password={db_password} host={db_host}"
+    try:
+        conn = psycopg2.connect(dsn)
+        conn.autocommit = True
+        return conn
+    except psycopg2.OperationalError as e:
+        logging.error("Failed to connect to database: %s", e)
+        return None
+
+
+def delete_fingerprints(
+    cursor: psycopg2.extensions.cursor, fingerprints: List[str], comment: str
+) -> None:
+    """Delete fingerprints from the database.
+
+    Args:
+        cursor: the database cursor.
+        fingerprints: list of fingerprints to delete.
+        comment: the comment associated with the deletion.
+    """
+    logging.info("Deleting fingerprints: %s", ", ".join(fingerprints))
+    try:
+        for fingerprint in fingerprints:
+            cursor.execute(
+                """
+                INSERT INTO deleted_keys (fingerprint, comment)
+                VALUES (LOWER(%s), %s)
+                ON CONFLICT DO NOTHING;
+            """,
+                (fingerprint, comment),
+            )
+
+        cursor.execute(
+            """
+            DELETE FROM subkeys USING deleted_keys
+            WHERE subkeys.rfingerprint = REVERSE(deleted_keys.fingerprint);
+        """
+        )
+        cursor.execute("ALTER TABLE subkeys DROP CONSTRAINT IF EXISTS subkeys_rfingerprint_fkey;")
```

Comment:
If you have sufficient permissions on the DB you can:
`SET session_replication_role = 'replica'` instead of removing/creating the constraint
I think there's also another approach to defer the constraints to the moment the commit is sent instead of when the question is executed